pkg/gadgets/socket-collector/tracer/tracer.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+package tracer
+
+import (
+	"bufio"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"net"
+	"runtime"
+
+	"github.com/cilium/ebpf/link"
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/api/v1alpha1"
+	"github.com/vishvananda/netns"
+)
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -cc clang IterTCPv4 ./bpf/tcp4-collector.c -- -I../../.. -O2 -g -c -x c
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -cc clang IterUDPv4 ./bpf/udp4-collector.c -- -I../../.. -O2 -g -c -x c
+
+type Socket struct {
+	Protocol            string `json:"protocol"`
+	LocalAddress        string `json:"local_address"`
+	LocalPort           uint16 `json:"local_port"`
+	RemoteAddress       string `json:"remote_address"`
+	RemotePort          uint16 `json:"remote_port"`
+	Status              string `json:"status"`
+	KubernetesNamespace string `json:"kubernetes_namespace"`
+	KubernetesPod       string `json:"kubernetes_pod"`
+}
+
+func parseIPv4(ipU32 uint32) string {
+	ipBytes := make([]byte, 4)
+
+	binary.BigEndian.PutUint32(ipBytes, ipU32)
+	ip := net.IP(ipBytes)
+
+	return ip.String()
+}
+
+// Parse is done based on what is defined kernel: https://github.com/torvalds/linux/blob/v5.13/include/net/tcp_states.h#L12-L24
+func parseStatus(proto string, statusUint uint8) (string, error) {
+	statusMap := [...]string{"ESTABLISHED", "SYN_SENT", "SYN_RECV", "FIN_WAIT1", "FIN_WAIT2", "TIME_WAIT", "CLOSE", "CLOSE_WAIT", "LAST_ACK", "LISTEN", "CLOSING", "NEW_SYN_RECV"}
+
+	// Kernel enum starts from 1, adjust it to the statusMap
+	statusUint--
```

Comment:
What would happen if `statusUint` is `0`?

Reply:
It's a `uint8` so after `--` operation it would be `255` which would make `if len(statusMap) <= int(statusUint)` statement true and return with an error as expected. However, I am not contrary to make it less implicit and add an if for `0` case. What do you think?

Reply:
I'd avoid relying on the overflow there. I think it's much cleaner to just compare against 0.

Reply:
Done